File: tutorials/csharp/pub_sub/ps_server/hwImpl.cs
/* **** including all of defines, service id(s) and request id(s) ***** */
using System;
using SocketProAdapter;
using SocketProAdapter.ServerSide;

//server implementation for service HelloWorld
public class HelloWorldPeer : CClientPeer
{
    public static string ToString(uint[] groups)
    {
        int n = 0;
        string s = "[";
        foreach (uint id in groups)
        {
            if (n != 0)
                s += ", ";
            s += id;
            ++n;
        }
        s += "]";
        return s;
    }

    protected override void OnSwitchFrom(uint oldServiceId)
    {
        uint[] chat_groups = { 1, 3 };
        Push.Subscribe(chat_groups);
    }

    protected override void OnSubscribe(uint[] groups)
    {
        Console.WriteLine(UID + " subscribes for groups " + ToString(groups));
    }

    protected override void OnUnsubscribe(uint[] groups)
    {
        Console.WriteLine(UID + " unsubscribes for groups " + ToString(groups));
    }

    protected override void OnPublish(object message, uint[] groups)
    {
        Console.WriteLine(UID + " publishes a message (" + message + ") to groups " + ToString(groups));
    }

    protected override void OnSendUserMessage(string receiver, object message)
    {
        Console.WriteLine(UID + " sends a message (" + message + ") to " + receiver);
    }

	[RequestAttr(hwConst.idSayHelloHelloWorld)]
	private string SayHello(string firstName, string lastName)
	{
        //processed within main thread
        System.Diagnostics.Debug.Assert(CSocketProServer.IsMainThread);

        //notify a message to groups [2, 3] at server side
        Push.Publish("Say hello from " + firstName + " " + lastName, 2, 3);

        string res = "Hello " + firstName + " " + lastName;
        Console.WriteLine(res);
        return res;
	}

	[RequestAttr(hwConst.idSleepHelloWorld, true)] //true -- slow request
	private void Sleep(int ms)
	{
        //processed within a worker thread
        System.Diagnostics.Debug.Assert(!CSocketProServer.IsMainThread);

        System.Threading.Thread.Sleep(ms);
	}

    [RequestAttr(hwConst.idEchoHelloWorld)]
    private CMyStruct Echo(CMyStruct ms)
    {
        return ms;
    }
}

File: tutorials/vbnet/pub_sub/ps_server/hwImpl.vb
' **** including all of defines, service id(s) and request id(s) ***** 
Imports SocketProAdapter.ServerSide

'server implementation for service HelloWorld
Public Class HelloWorldPeer
	Inherits CClientPeer

	Public Shared Overloads Function ToString(ByVal groups() As UInteger) As String
		Dim n As Integer = 0
		Dim s As String = "["
		For Each id As UInteger In groups
			If n <> 0 Then
				s &= ", "
			End If
			s &= id
			n += 1
		Next id
		s &= "]"
		Return s
	End Function

	Protected Overrides Sub OnSwitchFrom(ByVal oldServiceId As UInteger)
		Dim chat_groups() As UInteger = { 1, 3 }
		Push.Subscribe(chat_groups)
	End Sub

	Protected Overrides Sub OnSubscribe(ByVal groups() As UInteger)
		Console.WriteLine(UID & " subscribes for groups " & ToString(groups))
	End Sub

	Protected Overrides Sub OnUnsubscribe(ByVal groups() As UInteger)
		Console.WriteLine(UID & " unsubscribes for groups " & ToString(groups))
	End Sub

	Protected Overrides Sub OnPublish(ByVal message As Object, ByVal groups() As UInteger)
        Console.WriteLine(UID & " publishes a message (" & message.ToString() & ") to groups " & ToString(groups))
	End Sub

	Protected Overrides Sub OnSendUserMessage(ByVal receiver As String, ByVal message As Object)
        Console.WriteLine(UID & " sends a message (" & message.ToString() & ") to " & receiver)
	End Sub

	<RequestAttr(hwConst.idSayHelloHelloWorld)>
	Private Function SayHello(ByVal firstName As String, ByVal lastName As String) As String
		'processed within main thread
		System.Diagnostics.Debug.Assert(CSocketProServer.IsMainThread)

		'notify a message to groups [2, 3] at server side
		Push.Publish("Say hello from " & firstName & " " & lastName, 2, 3)

		Dim res As String = "Hello " & firstName & " " & lastName
		Console.WriteLine(res)
		Return res
	End Function

	<RequestAttr(hwConst.idSleepHelloWorld, True)>
	Private Sub Sleep(ByVal ms As Integer) 'true -- slow request
		'processed within a worker thread
		System.Diagnostics.Debug.Assert(Not CSocketProServer.IsMainThread)

		System.Threading.Thread.Sleep(ms)
	End Sub

	<RequestAttr(hwConst.idEchoHelloWorld)>
	Private Function Echo(ByVal ms As CMyStruct) As CMyStruct
		Return ms
	End Function
End Class

